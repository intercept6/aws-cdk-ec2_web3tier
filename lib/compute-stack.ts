import cdk = require('@aws-cdk/core');
import {AutoScalingGroup} from "@aws-cdk/aws-autoscaling";
import {
    AmazonLinuxImage,
    InstanceClass,
    InstanceSize,
    InstanceType,
    ISecurityGroup,
    IVpc,
    SubnetType
} from "@aws-cdk/aws-ec2";
import {ApplicationLoadBalancer} from "@aws-cdk/aws-elasticloadbalancingv2";
import {Role} from "@aws-cdk/aws-iam";

interface ComputeStackProps extends cdk.StackProps {
    vpc: IVpc;
    instanceRole: Role;
    appSg: ISecurityGroup;
}

export class ComputeStack extends cdk.Stack {
    constructor(scope: cdk.Construct, id: string, props: ComputeStackProps) {
        super(scope, id, props);


        const prj: string = this.node.tryGetContext('prj');
        const stage: string = this.node.tryGetContext('stage');
        // const params: any = this.node.tryGetContext(stage);

        const asg = new AutoScalingGroup(this, 'ASG', {
            vpc: props.vpc,
            instanceType: InstanceType.of(InstanceClass.BURSTABLE2, InstanceSize.MICRO),
            machineImage: new AmazonLinuxImage(),
            allowAllOutbound: true,
            role: props.instanceRole
        });
        asg.addUserData(
            "yum -y update",
            "yum -y install nginx",
            "systemctl enable nginx",
            "systemctl start nginx"
        );

        const alb = new ApplicationLoadBalancer(this, 'Alb', {
            vpc: props.vpc,
            vpcSubnets: {subnetType: SubnetType.PUBLIC},
            internetFacing: true,
            loadBalancerName: `${prj}-${stage}-alb`
        });

        const listener = alb.addListener('Listener', {
            port: 80,
        });
        listener.addTargets('Target', {
            targets: [asg],
            port: 80
        });

    }
}