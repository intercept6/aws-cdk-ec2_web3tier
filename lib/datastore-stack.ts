// @aws-cdk/aws-rds is a developer preview (public beta) module.
// Releases might lack important features and might have future breaking changes.

import {
  Construct,
  RemovalPolicy,
  Stack,
  StackProps,
  Tag
} from "@aws-cdk/core";
import {
  InstanceClass,
  InstanceSize,
  InstanceType,
  ISecurityGroup,
  IVpc,
  SubnetType
} from "@aws-cdk/aws-ec2";
import { DatabaseCluster, DatabaseClusterEngine } from "@aws-cdk/aws-rds";

interface DataStoreStackProps extends StackProps {
  vpc: IVpc;
  dbSg: ISecurityGroup;
}

export class DataStoreStack extends Stack {
  constructor(scope: Construct, id: string, props: DataStoreStackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: any = this.node.tryGetContext(stage);

    const cluster = new DatabaseCluster(this, "DBCluster", {
      engine: DatabaseClusterEngine.AURORA,
      masterUser: {
        username: "admin"
      },
      removalPolicy: RemovalPolicy.DESTROY,
      instanceProps: {
        instanceType: InstanceType.of(
          InstanceClass.MEMORY5,
          InstanceSize.LARGE
        ),
        vpcSubnets: {
          subnetType: SubnetType.ISOLATED
        },
        vpc: props.vpc,
        securityGroup: props.dbSg
      }
    });
    cluster.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
