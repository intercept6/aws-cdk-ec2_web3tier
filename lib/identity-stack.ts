import { Construct, Stack, StackProps } from "@aws-cdk/core";
import { ManagedPolicy, Role, ServicePrincipal } from "@aws-cdk/aws-iam";

export class IdentityStack extends Stack {
  public readonly instanceRole: Role;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const prj: string = this.node.tryGetContext("prj");
    const stage: string = this.node.tryGetContext("stage");
    // const params: object = this.node.tryGetContext(stage);

    this.instanceRole = new Role(this, "IamRole", {
      assumedBy: new ServicePrincipal("ec2.amazonaws.com"),
      roleName: `${prj}-${stage}-app`,
      managedPolicies: [
        ManagedPolicy.fromAwsManagedPolicyName(
          "service-role/AmazonEC2RoleforSSM"
        )
      ]
    });
  }
}
