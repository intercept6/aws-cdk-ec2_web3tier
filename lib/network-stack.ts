import { Construct, Stack, StackProps, Tag } from "@aws-cdk/core";
import { Port, SecurityGroup, SubnetType, Vpc } from "@aws-cdk/aws-ec2";

export class NetworkStack extends Stack {
    public readonly vpc: Vpc;
    public readonly appSg: SecurityGroup;
    public readonly dbSg: SecurityGroup;

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        const prj: string = this.node.tryGetContext("prj");
        const stage: string = this.node.tryGetContext("stage");
        const params: any = this.node.tryGetContext(stage);

    this.vpc = new Vpc(this, "VPC", {
      enableDnsHostnames: true,
      enableDnsSupport: true,
      natGateways: params.nat_gateways,
      maxAzs: 2,
      subnetConfiguration: [
        { name: "Public", subnetType: SubnetType.PUBLIC, cidrMask: 24 },
        { name: "Private", subnetType: SubnetType.PRIVATE, cidrMask: 24 },
        { name: "Isolated", subnetType: SubnetType.ISOLATED, cidrMask: 24 }
      ]
    });
    this.vpc.node.applyAspect(new Tag("Name", `${prj}-${stage}-vpc`));

    this.appSg = new SecurityGroup(this, "AppSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-app`,
      description: `${prj}-${stage}-app`
    });
    this.appSg.addIngressRule(
      this.appSg,
      Port.tcp(80),
      "allow http access from alb"
    );
    this.appSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-app`));

    this.dbSg = new SecurityGroup(this, "DbSg", {
      allowAllOutbound: true,
      vpc: this.vpc,
      securityGroupName: `${prj}-${stage}-db`,
      description: `${prj}-${stage}-db`
    });
    this.dbSg.addIngressRule(
      this.appSg,
      Port.tcp(3306),
      "allow db access from app server"
    );
    this.dbSg.node.applyAspect(new Tag("Name", `${prj}-${stage}-db`));
  }
}
