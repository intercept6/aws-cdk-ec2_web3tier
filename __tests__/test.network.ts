import {App} from "@aws-cdk/core"
import {SynthUtils} from "@aws-cdk/assert"
import {NetworkStack} from '../lib/network-stack'

describe("network", () => {
    test("default", () => {
        const app = new App();
        app.node.setContext("prj", "ssprj");
        app.node.setContext("stage", "ssstage");
        app.node.setContext("ssstage", {
            "nat_gateways": 1
        });
        const stack = new NetworkStack(app, 'TestNetworkStack');
        expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
    })
});