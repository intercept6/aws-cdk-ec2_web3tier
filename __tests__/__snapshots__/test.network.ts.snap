// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`network default 1`] = `
Object {
  "Resources": Object {
    "AppSg898BCA4E": Object {
      "Properties": Object {
        "GroupDescription": "ssprj-ssstage-app",
        "GroupName": "ssprj-ssstage-app",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-app",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "AppSgfromTestNetworkStackAppSg7494E757803596F2FC": Object {
      "Properties": Object {
        "Description": "allow http access from alb",
        "FromPort": 80,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "AppSg898BCA4E",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "AppSg898BCA4E",
            "GroupId",
          ],
        },
        "ToPort": 80,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "DbSg495F8305": Object {
      "Properties": Object {
        "GroupDescription": "ssprj-ssstage-db",
        "GroupName": "ssprj-ssstage-db",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-db",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "DbSgfromTestNetworkStackAppSg7494E7573306D05F2807": Object {
      "Properties": Object {
        "Description": "allow db access from app server",
        "FromPort": 3306,
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "DbSg495F8305",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "AppSg898BCA4E",
            "GroupId",
          ],
        },
        "ToPort": 3306,
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "VPCB9E5F0B4": Object {
      "Properties": Object {
        "CidrBlock": "10.0.0.0/16",
        "EnableDnsHostnames": true,
        "EnableDnsSupport": true,
        "InstanceTenancy": "default",
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-vpc",
          },
        ],
      },
      "Type": "AWS::EC2::VPC",
    },
    "VPCIGWB7E252D3": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "ssprj-ssstage-vpc",
          },
        ],
      },
      "Type": "AWS::EC2::InternetGateway",
    },
    "VPCIsolatedSubnet1RouteTableAssociationA2D18F7C": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCIsolatedSubnet1RouteTableEB156210",
        },
        "SubnetId": Object {
          "Ref": "VPCIsolatedSubnet1SubnetEBD00FC6",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCIsolatedSubnet1RouteTableEB156210": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/IsolatedSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCIsolatedSubnet1SubnetEBD00FC6": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.4.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/IsolatedSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCIsolatedSubnet2RouteTable9B4F78DC": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/IsolatedSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCIsolatedSubnet2RouteTableAssociation7BF8E0EB": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCIsolatedSubnet2RouteTable9B4F78DC",
        },
        "SubnetId": Object {
          "Ref": "VPCIsolatedSubnet2Subnet4B1C8CAA",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCIsolatedSubnet2Subnet4B1C8CAA": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.5.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/IsolatedSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Isolated",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Isolated",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPrivateSubnet1DefaultRouteAE1D6490": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VPCPublicSubnet1NATGatewayE0556630",
        },
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet1RouteTableBE8A6027",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPrivateSubnet1RouteTableAssociation347902D1": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet1RouteTableBE8A6027",
        },
        "SubnetId": Object {
          "Ref": "VPCPrivateSubnet1Subnet8BCA10E0",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPrivateSubnet1RouteTableBE8A6027": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PrivateSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPrivateSubnet1Subnet8BCA10E0": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.2.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PrivateSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPrivateSubnet2DefaultRouteF4F5CFD2": Object {
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "NatGatewayId": Object {
          "Ref": "VPCPublicSubnet1NATGatewayE0556630",
        },
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet2RouteTable0A19E10E",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPrivateSubnet2RouteTable0A19E10E": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PrivateSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPrivateSubnet2RouteTableAssociation0C73D413": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPrivateSubnet2RouteTable0A19E10E",
        },
        "SubnetId": Object {
          "Ref": "VPCPrivateSubnet2SubnetCFCDAA7A",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPrivateSubnet2SubnetCFCDAA7A": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.3.0/24",
        "MapPublicIpOnLaunch": false,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PrivateSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Private",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Private",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPublicSubnet1DefaultRoute91CEF279": Object {
      "DependsOn": Array [
        "VPCVPCGW99B986DC",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet1RouteTableFEE4B781",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPublicSubnet1EIP6AD938E8": Object {
      "Properties": Object {
        "Domain": "vpc",
      },
      "Type": "AWS::EC2::EIP",
    },
    "VPCPublicSubnet1NATGatewayE0556630": Object {
      "Properties": Object {
        "AllocationId": Object {
          "Fn::GetAtt": Array [
            "VPCPublicSubnet1EIP6AD938E8",
            "AllocationId",
          ],
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet1SubnetB4246D30",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PublicSubnet1",
          },
        ],
      },
      "Type": "AWS::EC2::NatGateway",
    },
    "VPCPublicSubnet1RouteTableAssociation0B0896DC": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet1RouteTableFEE4B781",
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet1SubnetB4246D30",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPublicSubnet1RouteTableFEE4B781": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PublicSubnet1",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPublicSubnet1SubnetB4246D30": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            0,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.0.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PublicSubnet1",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCPublicSubnet2DefaultRouteB7481BBA": Object {
      "DependsOn": Array [
        "VPCVPCGW99B986DC",
      ],
      "Properties": Object {
        "DestinationCidrBlock": "0.0.0.0/0",
        "GatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet2RouteTable6F1A15F1",
        },
      },
      "Type": "AWS::EC2::Route",
    },
    "VPCPublicSubnet2RouteTable6F1A15F1": Object {
      "Properties": Object {
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PublicSubnet2",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::RouteTable",
    },
    "VPCPublicSubnet2RouteTableAssociation5A808732": Object {
      "Properties": Object {
        "RouteTableId": Object {
          "Ref": "VPCPublicSubnet2RouteTable6F1A15F1",
        },
        "SubnetId": Object {
          "Ref": "VPCPublicSubnet2Subnet74179F39",
        },
      },
      "Type": "AWS::EC2::SubnetRouteTableAssociation",
    },
    "VPCPublicSubnet2Subnet74179F39": Object {
      "Properties": Object {
        "AvailabilityZone": Object {
          "Fn::Select": Array [
            1,
            Object {
              "Fn::GetAZs": "",
            },
          ],
        },
        "CidrBlock": "10.0.1.0/24",
        "MapPublicIpOnLaunch": true,
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "TestNetworkStack/VPC/PublicSubnet2",
          },
          Object {
            "Key": "aws-cdk:subnet-name",
            "Value": "Public",
          },
          Object {
            "Key": "aws-cdk:subnet-type",
            "Value": "Public",
          },
        ],
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::Subnet",
    },
    "VPCVPCGW99B986DC": Object {
      "Properties": Object {
        "InternetGatewayId": Object {
          "Ref": "VPCIGWB7E252D3",
        },
        "VpcId": Object {
          "Ref": "VPCB9E5F0B4",
        },
      },
      "Type": "AWS::EC2::VPCGatewayAttachment",
    },
  },
}
`;
